tools/autograph-monitor/monitor.go

```diff
@@ -0,0 +1,361 @@
+		certs, err = getX5U(sr.X5U)
+		if err != nil {
+			return err
+		}
+		if len(certs) < 2 {
+			return fmt.Errorf("Found %d certs in X5U, expected at least 2", len(certs))
+		}
+		// certs[0] is the end entity
+		pubkey = certs[0].PublicKey.(*ecdsa.PublicKey)
+	} else {
+		pubkey, err = parsePublicKeyFromB64(sr.PublicKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	if sr.ContentSignature != "" {
+		data = make([]byte, len("Content-Signature:\x00")+len(inputdata))
+		copy(data[:len("Content-Signature:\x00")], []byte("Content-Signature:\x00"))
+		copy(data[len("Content-Signature:\x00"):], inputdata)
+	} else {
+		data = make([]byte, len(inputdata))
+		copy(data, inputdata)
+	}
+	datahash, err := digest(data, sr.Hash)
+	if err != nil {
+		return err
+	}
+	sigBytes, err := fromBase64URL(sr.Signature)
+	if err != nil {
+		return err
+	}
+	r, s := new(big.Int), new(big.Int)
+	r.SetBytes(sigBytes[:len(sigBytes)/2])
+	s.SetBytes(sigBytes[len(sigBytes)/2:])
+	if !ecdsa.Verify(pubkey, datahash, r, s) {
+		return fmt.Errorf("Signature verification failed")
+	}
+
+	if certs != nil {
+		return verifyCertChain(certs)
+	}
+
+	return nil
+}
+
+func getX5U(x5u string) (certs []*x509.Certificate, err error) {
+	resp, err := http.Get(x5u)
+	if err != nil {
+		return certs, fmt.Errorf("Failed to retrieve X5U %s: %v", x5u, err)
+	}
+	defer resp.Body.Close()
+	scanner := bufio.NewScanner(resp.Body)
+	// the first row must contain BEGIN CERT for the end entity
+	scanner.Scan()
+	if scanner.Text() != "-----BEGIN CERTIFICATE-----" {
+		return certs, fmt.Errorf("Invalid X5U format for %s: first row isn't BEGIN CERTIFICATE", x5u)
+	}
+	var certPEM []byte
+	certPEM = append(certPEM, scanner.Bytes()...)
+	certPEM = append(certPEM, byte('\n'))
+	for scanner.Scan() {
+		certPEM = append(certPEM, scanner.Bytes()...)
+		certPEM = append(certPEM, byte('\n'))
+		if scanner.Text() == "-----END CERTIFICATE-----" {
+			// end of the current cert. Parse it, store it
+			// and move on to next cert
+			block, _ := pem.Decode(certPEM)
+			if block == nil {
+				return certs, fmt.Errorf("Failed to parse certificate PEM")
+			}
+			certX509, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return certs, fmt.Errorf("Could not parse X.509 certificate: %v", err)
+
+			}
+			certs = append(certs, certX509)
+			certPEM = nil
+		}
+	}
+	return certs, nil
+}
+
+func parsePublicKeyFromB64(b64PubKey string) (pubkey *ecdsa.PublicKey, err error) {
+	keyBytes, err := fromBase64URL(b64PubKey)
+	if err != nil {
+		return pubkey, fmt.Errorf("Failed to parse public key base64: %v", err)
+	}
+	keyInterface, err := x509.ParsePKIXPublicKey(keyBytes)
+	if err != nil {
+		return pubkey, fmt.Errorf("Failed to parse public key DER: %v", err)
+	}
+	pubkey = keyInterface.(*ecdsa.PublicKey)
+	return pubkey, nil
+}
+
+func verifyCertChain(certs []*x509.Certificate) error {
+	for i, cert := range certs {
+		if (i + 1) == len(certs) {
+			// this is the last cert, it should be self signed
```

Comment:
You can't rely on the order of certs in the chain (well you can because you know what you get, but in general you can't).